spirv_glsl.cpp

```diff
@@ -8889,30 +8889,35 @@ string CompilerGLSL::access_chain_internal(uint32_t base, const uint32_t *indice
 				                                       is_packed);
 			}
 
-			if (is_literal && !is_packed && !row_major_matrix_needs_conversion)
+			if (is_literal)
 			{
-				expr += ".";
-				expr += index_to_swizzle(index);
+				if (!is_packed && !row_major_matrix_needs_conversion)
+				{
+					expr += ".";
+					expr += index_to_swizzle(index);
+				}
+				else
+				{
+					// For packed vectors, we can only access them as an array, not by swizzle.
+					expr += join("[", index, "]");
+				}
 			}
 			else if (ir.ids[index].get_type() == TypeConstant && !is_packed && !row_major_matrix_needs_conversion)
 			{
 				auto &c = get<SPIRConstant>(index);
+				bool out_of_bounds = (c.scalar() >= type->vecsize);
```

Comment:
This can get spicy with OpSpecConstantOp. In that we cannot directly look at the scalar value, but the MSL compiler might still complain ... You might want to add a test for that and consider looking at `Compiler::evaluate_spec_constant_u32()`.

Reply:
Or hm ... we won't know at compile time what the result is going to be. Perhaps do a dynamic "out_of_bounds" check on the spec constant op, e.g.

```
foo[spec_constant_op_result >= size ? 0 : spec_constant_op_result]
```

Reply:
It seems like the `OpSpecConstantOp` is of type `TypeConstantOp` so it would be excluded of that branch specifically (it would fall in the `else` condition).

I was targeting the `OpAccessChain` mainly, but now that I think about it, when unrolling the indexes there's no check for `OpAccessChain` or `OpInBoundsAccessChain`.  I would say that adding the dynamic check would be necessary only in the latter but I'm willing to change it too.

Let me know what you think!

Reply:
Wouldn't it be necessary in both? InBoundsAccessChain guarantees that it's in bound, but I guess only if it's executed.